metatensor-torch/src/atomistic.cpp

```diff
@@ -111,71 +113,81 @@ NeighborsListOptions NeighborsListOptionsHolder::from_json(const std::string& js
-            "invalid samples for `positions`: the samples names must be "
-            "'atom' and 'species'"
+            "`species`, `positions`, and `cell` must be on the same device, got " +
+            species_.device().str() + ", " + positions_.device().str() + ", and " +
+            cell_.device().str()
         );
     }
 
-    auto components = positions->components();
-    if (components.size() != 1 || *components[0] != metatensor::Labels({"xyz"}, {{0}, {1}, {2}})) {
+    if (species_.sizes().size() != 1) {
         C10_THROW_ERROR(ValueError,
-            "invalid components for `positions`: there should be a single 'xyz'=[0, 1, 2] component"
+            "`species` must be a 1 dimensional tensor, got a tensor with " +
+            std::to_string(species_.sizes().size()) + " dimensions"
         );
     }
 
-    if (*positions->properties() != metatensor::Labels({"position"}, {{0}})) {
+    if (torch::canCast(species_.scalar_type(), torch::kInt32)) {
+        species_ = species_.to(torch::kInt32);
+    } else {
         C10_THROW_ERROR(ValueError,
-            "invalid properties for `positions`: there should be a single 'positions'=0 property"
+            "`species` must be a tensor of integers, got " +
+            scalar_type_name(species_.scalar_type()) + " instead"
         );
     }
 
-    if (!positions->gradients_list().empty()) {
-        C10_THROW_ERROR(ValueError, "`positions` should not have any gradients");
-    }
-
-    // check the cell
-    if (*cell->samples() != metatensor::Labels({"_"}, {{0}})) {
+    auto n_atoms = species_.size(0);
+    if (positions_.sizes().size() != 2) {
         C10_THROW_ERROR(ValueError,
-            "invalid samples for `cell`: there should be a single '_'=0 sample"
+            "`positions` must be a 2 dimensional tensor, got a tensor with " +
+            std::to_string(positions_.sizes().size()) + " dimensions"
         );
     }
 
-    components = cell->components();
-    if (components.size() != 2) {
+    if (positions_.size(0) != n_atoms || positions_.size(1) != 3) {
         C10_THROW_ERROR(ValueError,
-            "invalid components for `cell`: there should be 2 components, got "
-            + std::to_string(components.size())
+            "`positions` must be a (n_atoms x 3) tensor, got a tensor with shape [" +
+            std::to_string(positions_.size(0)) + ", " + std::to_string(positions_.size(1)) + "]"
         );
     }
 
-    if (*components[0] != metatensor::Labels({"cell_abc"}, {{0}, {1}, {2}})) {
+    if (!is_floating_point(positions_.scalar_type())) {
         C10_THROW_ERROR(ValueError,
-            "invalid components for `cell`: the first component should be 'cell_abc'=[0, 1, 2]"
+            "`positions` must be a tensor of floating point data, got " +
+            scalar_type_name(positions_.scalar_type()) + " instead"
         );
     }
 
-    if (*components[1] != metatensor::Labels({"xyz"}, {{0}, {1}, {2}})) {
+    if (cell_.sizes().size() != 2) {
```

Comment:
maybe explicitly check for  a 3x3 matrix here?

Reply:
just below

Reply:
ahh yes 🙈